bench/group.bench.ts

```diff
@@ -21,8 +21,8 @@ export async function benchGroup(bs: Benchmark.Suite) {
     const msg = te.encode('msg')
     const dst = te.encode('dst')
 
-    for (const id of Object.values(Oprf.Group.ID)) {
-        const gg = new Oprf.Group(id)
+    for (const id of Oprf.Group.supportedIDs) {
+        const gg = Oprf.Group.fromID(id)
```

Comment:
@armfazh 

In practice it seems like suite and group id are essentially the same? 
The GroupID largely used to reference a curve in sjcl ? 

If a "Group" has members like the following:
```
    hashToGroup(msg: Uint8Array, dst: Uint8Array): Promise<Elt>

    hashToScalar(msg: Uint8Array, dst: Uint8Array): Promise<Scalar>
```

Perhaps a Group should really be defined by its suite id?

Reply:
SuiteID is an identifier for a combination of algorithms (groupId,hashFn, ...), eg. SuiteABC:=(P-256,SHA-256).
In the suites currently supported there is a 1-to-1 correspondence between suiteID and Groups, which may hint that they are the same.

However, nothing restricts the definition of a new suite such as SuiteXYZ:=(P-256,SHA-512).

So SuiteID identifies an (V)OPRF instance, while the GroupID identifies a specific (elliptic curve) group.

Reply:
#### Background

In the current codebase, there exists a TypeScript interface named `Group` designed to encapsulate various cryptographic operations. Specifically, it combines elliptic curve operations, like addition and multiplication over a particular curve, with hash functions (`hashToGroup`, `hashToScalar`). 

However, this `Group` interface also does more than just representing a mathematical group; it effectively combines a curve with a set of hashing operations, thus acting as a combination of cryptographic primitives, commonly referred to as a "suite" in cryptography.

#### Issue 1: Semantic Misalignment

The naming of the `Group` interface doesn't fully convey the suite-like nature of what it represents, potentially leading to semantic misunderstandings. In cryptographic terms, a "suite" usually consists of a combination of cryptographic algorithms and parameters, which matches the current design of the `Group` interface more closely.

#### Issue 2: Redundant Identifiers

Additionally, there are two types of identifiers used to represent these configurations: `GroupID` and `SuiteID`. The presence of both can cause redundancy and confusion when reading or working with the code.

#### Issue 3: Conflating Group and Hash Operations

The current `Group` interface bundles elliptic curve operations and hash functions together, combining two sets of functionalities that could be separate concerns.

#### Considerations

1. Renaming `Group` to `Suite` would require extensive refactoring that touches many parts of the code, tests, and potentially external documentation.
2. There's an option to unify `GroupID` and `SuiteID`, either by making one an alias for the other or by collapsing them into a single, new identifier.
3. A less invasive alternative is to retain the `Group` name but clarify its dual nature through documentation comments.
4. Separating these concerns into distinct interfaces would make the responsibilities of each clearer but could lead to additional complexity, particularly if the `Group` term is pervasively used to refer to what would essentially be a `CryptoSuite`.

#### Possible Solutions

1. **Full Refactor**: Rename `Group` to `Suite`, and replace all instances. Unify `GroupID` and `SuiteID`.
   
2. **Partial Refactor**: Keep `Group` but make `GroupID` an alias for `SuiteID`. Add comprehensive documentation to clarify the nature of the `Group` interface.

3. **Minimal Change**: Retain all current naming but add extensive documentation to clarify the terms.

4. **Separate Concerns**: Divide the existing `Group` interface into two interfaces:

   - `Group`: Includes only the elliptic curve operations.
   - `HashSuite` or similar: Includes the hash-to-curve methods (`hashToGroup`, `hashToScalar`).

   Each `Group` could then be paired with a `HashSuite`, either through composition or another design pattern, to form a complete cryptographic suite.

The choice among these options hinges on balancing the semantic clarity gained through renaming against the practical costs of extensive code changes.

Reply:
#### Recap

1. **Identification of GroupID**: Initially, `GroupID` was observed to identify only the curve used, yet it implicitly determined the choice of hash. This was the first sign of semantic misalignment.

2. **Expanding GroupID**: To make the implication explicit, the hash choice was added to each `GroupID` constant, signaling that it wasn't just about the curve.

3. **Redundancy Observed**: At this point, it became clear that `GroupID` and `SuiteID` were effectively doing the same thing: identifying a cryptographic suite.

4. **Initial Unification**: An alias was created where `SuiteID` was used for `GroupID`, eliminating redundancy to some extent.

5. **Complete Unification**: The next logical step was to remove redundancy further by simply using `SuiteID` everywhere, streamlining the identifiers.

6. **A Semantic Dilemma**: Despite the unification, it became awkward that a method named `fromID` on a `Group` would take a `SuiteID`. This surfaced the initial issue of semantic misalignment more prominently.

7. **Full Renaming**: Finally, to align with the semantically correct cryptographic terminology, `Group` was renamed to `Suite` across the codebase.

#### Resolved Issues

1. **Semantic Misalignment**: Renaming `Group` to `Suite` resolved the initial semantic inconsistency by aligning the name with its actual function as a cryptographic suite.

2. **Redundant Identifiers**: Unifying `GroupID` and `SuiteID` into a single `SuiteID` eliminated the redundancy and simplified the codebase.

#### Final Thoughts

The entire process was a sequence of incremental realizations, each leading to more streamlined and semantically accurate code. However, this did require a significant amount of refactoring, given the pervasive use of the term `Group`.  This also unfortunately meant breaking the previously exported interface. 

It's not clear to me what's the ideal way to address this is.

Reply:
Ok, so my confusion seems to be around the `hash` 
The Group "explicitly implies" a given hash, but that hash is separate to the CipherSuite's choice of hash
(which just happens to be the same here) 

In other words, you can't "configure" the Group hash for the hashToGroup/hashToScalar functions

Reply:
I guess GroupID is essentially referring to a hash-to-curve "suite":
https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16#section-8.3

Reply:
```
Since hash_to_ristretto255 is not a hash-to-curve suite, it does not
   have a Suite ID.  If a similar identifier is needed, it MUST be
   constructed following the guidelines in [Section 8.10](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16#section-8.10), with the
   following parameters:

   *  CURVE_ID: "ristretto255"

   *  HASH_ID: as described in [Section 8.10](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16#section-8.10)

   *  MAP_ID: "R255MAP"

   *  ENC_VAR: "RO"

   For example, if expand_message is expand_message_xmd using SHA-512,
   the REQUIRED identifier is:

   ristretto255_XMD:SHA-512_R255MAP_RO_
```